### DISCUSSION TITLE
Waiting for work done on chrome

### DISCUSSION THREAD

**Discussion by BLaZeKiLL** [gfx-rs/wgpu#5122] [Q&A]
I have a for loop that submits a compute pass every iteration. On native, I can use 
```rust
gpu.device
   .poll(wgpu::Maintain::WaitForSubmissionIndex(submission_index))
   .panic_on_timeout();
```
So that I can wait in between iterations of the loops and the next work is submitted only after previous is complete

I have been porting over my app to wasm, now I know in browsers poll won't do anything so I tried using `on_submitted_work_done` as follows
```rust
let submission_index = gpu.queue.submit([encoder.finish()]);

let (sender, receiver) = flume::bounded(1);

log::info!("waiting");
gpu.queue.on_submitted_work_done(move || sender.send(()).unwrap());
log::info!("done");
// since this won't do anything on wasm, added a separate channel notifier
gpu.device
    .poll(wgpu::Maintain::WaitForSubmissionIndex(submission_index))
    .panic_on_timeout();

if let Ok(()) = receiver.recv_async().await {
    return Ok(());
} else {
    return Err(());
}
```

When running in Chrome I get a panic at `gpu.queue.on_submitted_work_done(move || sender.send(()).unwrap());` stating it's not implemented

according to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/GPUQueue/onSubmittedWorkDone) `on_submitted_work_done` was implemented in chrome 113, (I am running chrome 120). I am assuming wgpu has yet to implement this.

So how can I perform a wait now? or should I just wait for wgpu to catch up

**Comment by cwfitzgerald** [comment]
Yeah I came across this in #5012 and was surprised it was completely unimplemented.

**Reply by cwfitzgerald** [reply]
You actually should be getting a panic - do you have a console panic hook registered?

**Reply by BLaZeKiLL** [reply]
Yeah I am getting a panic, for now I have removed it and let the for loop submit all the work one after the other

**Reply by BLaZeKiLL** [reply]
I can always make a app progressive and have a map_async on  each iteration and wait for that
